### PR TITLE
feat: make volume optional for ionoscloud_server

### PR DESCRIPTION
**Pull request by cristiGuranIonos**
## Checklist

- [ ] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [ ] Tests added or updated
- [ ] Documentation updated
- [ ] Changelog updated and version incremented (label: upcoming release)
- [ ] Github Issue linked if any
- [ ] Jira task updated

### CODE REVIEW

File: ionoscloud/resource_server.go [hunk 1]
```diff
@@ -568,65 +568,62 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
...
 
-	// add volume object to serverReq
-	serverReq.Entities = &ionoscloud.ServerEntities{
-		Volumes: &ionoscloud.AttachedVolumes{
-			Items: &[]ionoscloud.Volume{
-				{
-					Properties: volume,
+		}
+		// add volume object to serverReq
+		serverReq.Entities = &ionoscloud.ServerEntities{
```
Comment (adeatcu-ionos) [inline ionoscloud/resource_server.go:617]: @cristiGuranIonos since `serverReq.Entities` is instantiated above: `571: serverReq.Entities = ionoscloud.NewServerEntities()` I think we can just populate the `Volumes` property here.

File: ionoscloud/resource_server.go [hunk 2]
```diff
@@ -182,7 +182,7 @@ func resourceServer() *schema.Resource {
 			},
 			"volume": {
 				Type:     schema.TypeList,
-				Required: true,
+				Optional: true,
```
Comment (adeatcu-ionos) [inline ionoscloud/resource_server.go:185]: @cristiGuranIonos shouldn't we add a `ForceNew: true` here?

File: ionoscloud/resource_server.go [hunk 3]
```diff
@@ -767,9 +764,12 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
 						return diags
 					}
 				}
-
-				volumeItems := serverReq.Entities.Volumes.Items
-				firstVolumeItem := (*volumeItems)[0]
+				var volumeItems *[]ionoscloud.Volume
+				var firstVolumeItem ionoscloud.Volume
+				if serverReq.Entities.Volumes != nil {
+					volumeItems = serverReq.Entities.Volumes.Items
```
Comment (adeatcu-ionos) [inline ionoscloud/resource_server.go:770]: @cristiGuranIonos shouldn't we also check if `serverReq.Entities.Volumes.Items != nil` and `serverReq.Entities.Volumes.Items` is not an empty list?

File: ionoscloud/resource_server.go [hunk 4]
```diff
@@ -568,65 +568,62 @@ func resourceServerCreate(ctx context.Context, d *schema.ResourceData, meta inte
...
-	if err != nil {
-		return diag.FromErr(err)
-	}
-
-	// get image and imageAlias
-	image, imageAlias, err := getImage(ctx, client, d, *volume)
 	if err != nil {
 		return diag.FromErr(err)
 	}
+	if volume.Type != nil && *volume.Type != "" {
```
Comment (adeatcu-ionos) [inline ionoscloud/resource_server.go:578]: @cristiGuranIonos can you also please add a test in which we create a server without a volume? Also maybe, if it's not to complex, we can add some simple edge cases.

**Review by adeatcu-ionos — CHANGES_REQUESTED** [review]

File: ionoscloud/resource_server.go (hunk 2 above)
Comment (cristiGuranIonos) [reply ionoscloud/resource_server.go:185]: it has inner fields that require forcenew. we don't want to do it for all fields of the volume.

File: ionoscloud/resource_server.go (hunk 3 above)
Comment (cristiGuranIonos) [reply ionoscloud/resource_server.go:768]: added a check below before instantiating the first volume. if volumes is != nil then most likely items is populated

File: ionoscloud/resource_server.go (hunk 1 above)
Comment (cristiGuranIonos) [reply ionoscloud/resource_server.go]: done

File: ionoscloud/resource_server_test.go [hunk 5]
```diff
@@ -61,8 +61,6 @@ func TestAccServerBasic(t *testing.T) {
 					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "hostname", constant.ServerTestHostname),
 					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "cores", "1"),
 					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "ram", "1024"),
-					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "availability_zone", "ZONE_1"),
```
Comment (adeatcu-ionos) [inline ionoscloud/resource_server_test.go:64]: @cristiGuranIonos why are we removing these checks?

File: ionoscloud/resource_server_test.go [hunk 6]
```diff
@@ -323,7 +337,25 @@ func TestAccServerNoBootVolumeBasic(t *testing.T) {
...
+					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "nic.0.name", "system"),
+					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "nic.0.dhcp", "true"),
+					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "nic.0.firewall_active", "true"),
+					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "nic.0.mac", constant.NicMac),
+					resource.TestCheckResourceAttrPair(constant.ServerResource+"."+constant.ServerTestResource, "nic.0.id", constant.ServerResource+"."+constant.ServerTestResource, "primary_nic"),
+					resource.TestCheckResourceAttr(constant.ServerResource+"."+constant.ServerTestResource, "nic.0.firewall_type", "INGRESS"),
+				),
+			},
+			{
+				Config: testAccCheckServerConfigemoveServer,
```
Comment (adeatcu-ionos) [inline ionoscloud/resource_server_test.go:358]: @cristiGuranIonos `testAccCheckServerConfigRemoveServer`

File: ionoscloud/resource_server_test.go [hunk 7]
```diff
@@ -1868,8 +1900,47 @@ resource ` + constant.ServerResource + ` ` + constant.ServerTestResource + ` {
     firewall_active = false
   }
 }`
+const testAccCheckServerConfigNoVolume = `
```
Comment (adeatcu-ionos) [inline ionoscloud/resource_server_test.go:1903]: @cristiGuranIonos I would change this to `testAccCheckServerNoInlineVolumeConfig`, I would emphasize the fact that the volume is not inline (we have an external volume in this configuration).

**Review by adeatcu-ionos — CHANGES_REQUESTED** [review]

File: ionoscloud/resource_server_test.go (hunk 5 above)
Comment (cristiGuranIonos) [reply ionoscloud/resource_server_test.go:64]: they cause issues if the av zone or the cpu family is not available. better to let the api decide, we test other things.

**Review by adeatcu-ionos — APPROVED** [review]
